feat(admin): Add the ability to search by product title under purchase history in admin

omeraydindev opened the pull request:
Resolves #655.

Adds product title filtering to admin purchase search, to quickly find purchases when customers only provide product names.

## New Feature: Product Title Search & Filtering

This enhancement adds product title search capabilities to the admin purchase search system.

### Key Features
- **Standalone Product Search**: Search purchases by product title directly from the admin search page
- **Combined Search**: Filter existing email-based search results by product title for precise targeting
- **Case-Insensitive Matching**: Finds products regardless of capitalization differences
- **Partial Matching**: Supports substring searches (e.g., "Script" finds "JavaScript Masterclass")

## Changes
- **Service**: Extended `AdminSearchService` with product title search functionality
- **Controller**: Enhanced purchase search to handle product title parameters
- **UI**: Added integrated search form with icon-based design matching admin patterns

## Screenshots

<img width="1499" height="870" alt="image" src="https://github.com/user-attachments/assets/2a3aabe3-3b24-426f-9c76-c202c4fadd15" />

<img width="1498" height="858" alt="image" src="https://github.com/user-attachments/assets/82483e73-2c15-4655-83ab-85bce3180976" />
 

## Testing
- Comprehensive coverage including controller, service, and feature tests. These cover standalone and combined search scenarios.

Output of `bundle exec rspec spec/services/admin_search_service_spec.rb spec/controllers/admin/search_controller_spec.rb spec/requests/admin/pages_spec.rb --format html --out search_in_purchases_tests.html`: (The new tests)
[search_in_purchases_tests.pdf](https://github.com/user-attachments/files/21844441/search_in_purchases_tests.pdf)

omeraydindev commented on app/controllers/admin/users/purchases_controller.rb:
```diff
@@ -0,0 +1,33 @@
+      else
+        @user = User.find_by(username: params[:user_id]) ||
+                User.find_by(id: params[:user_id])
+      end
```
Do you think this is needed? @sharang-d

omeraydindev replied:
This file no longer exists

omeraydindev commented:
Closed some comments from CodeRabbit as they seemed to be false positives

sharang-d commented on app/controllers/admin/users/purchases_controller.rb:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+class Admin::Users::PurchasesController < Admin::BaseController
```
I guess Sahil means showing the list on the same page as users/id. Not necessarily clubbing the data in the same endpoint/controller.

slavingia replied:
Yep, exactly just in the same visual place since most of the Support team spends their time on those pages already

omeraydindev replied:
My bad, got it.

What about something like this then? By default no purchase history is shown, but when the user searches for something, they get up to 10 results back (to not clutter up the user page that much):

<img width="1512" height="904" alt="Screenshot 2025-08-04 at 23 36 24" src="https://github.com/user-attachments/assets/99b2d21d-f1ad-468a-9082-bcb4ff4312fa" />

<img width="1512" height="904" alt="Screenshot 2025-08-04 at 23 36 40" src="https://github.com/user-attachments/assets/4b02471e-7636-4e13-b7c7-d73cf9db6903" />

<img width="1512" height="904" alt="Screenshot 2025-08-04 at 23 37 07" src="https://github.com/user-attachments/assets/928cc20b-9d03-4ad9-861c-fe67232b77e1" />

omeraydindev replied:
@slavingia Kindly let me know if this makes sense or if you have something else in mind UI/UX-wise

omeraydindev commented:
Moved the purchase history UI to the /admin/users/:id page, per @slavingia's feedback above. Updated the PR description with new app screenshots and the revised test suite.

The spec for this feature is passing (see screenshot), but getting the entire test suite green locally is proving to be a bit tricky. I’ll follow up here once I manage to resolve it.

omeraydindev commented:
Output of `bundle exec rspec spec/controllers/admin spec/requests/admin --format html --out rspec_results.html`

<img width="1509" height="896" alt="image" src="https://github.com/user-attachments/assets/2a7937e9-c046-4949-be3e-490122344ee7" />

omeraydindev commented:
@slavingia Sure, here you go:

[RSpec results.pdf](https://github.com/user-attachments/files/21661209/RSpec.results.pdf)

omeraydindev commented:
@slavingia As in, move new specs for this feature from `spec/controllers/admin/users_controller_spec.rb` to a new spec file?

omeraydindev commented:
@slavingia Added the e2e request spec

`bundle exec rspec spec/requests/admin/purchase_history_search_spec.rb --format html --out purchase_history_search_spec.html`:

<img width="1505" height="841" alt="image" src="https://github.com/user-attachments/assets/c8e751e4-e391-411f-a986-27df51e044fc" />

hchhabra reviewed (changes requested):
The changes here don't fulfill the requirement set in the issue I think, as the requirement is to further filter the search results (searched initially by customer's email) by product title. Not all customers have a user account, so having search based on user will not be that useful imo, this should be a filter on purchase search results page i.e. basically a search on customer email + product title. @michellelarney can confirm. @omeraydindev _a

<img width="828" height="98" alt="Screenshot 2025-08-12 at 1 35 42 PM" src="https://github.com/user-attachments/assets/6737c48f-7c4d-4a48-aab4-d47f136c8f41" />

michellelarney commented:
I confirm what Harbaksh mentioned is correct. It is to be a search based on customer email + product title, since not all buyers have a user account.

slavingia commented:
Closing but re-open when you have something merge-able. Also, note we've updated our guidelines to require an AI disclosure, video of feature, light/dark/mobile/desktop UX considered, and more.
